esphome/components/azure_iot_hub/azure_iot_hub.cpp

```diff
@@ -0,0 +1,353 @@
+      fingerprint_string++;
+    }
+  }
+
+  return idx == 20;
+}
+#endif
+
+void AzureIoTHub::setup() {
+  this->setup_controller();
+#ifdef ARDUINO_ARCH_ESP8266
+  this->wifi_client_ = new BearSSL::WiFiClientSecure();
+
+  if (this->ssl_fingerprint_supplied_) {
+    ESP_LOGD(TAG, "Using SSL with SHA1 fingerprint verification");
+    this->wifi_client_->setFingerprint(this->ssl_sha1_fingerprint_bytes_);
+  } else {
+    ESP_LOGD(TAG, "Using insecure SSL");
+    this->wifi_client_->setInsecure();
+  }
+  this->wifi_client_->setBufferSizes(512, 512);
+#endif
+
+#ifdef ARDUINO_ARCH_ESP32
+  this->wifi_client_ = new WiFiClientSecure();
+  if (baltimore_root_pem_ != nullptr) {
+    ESP_LOGD(TAG, "Using SSL with Baltimore root certificate for verification");
+    this->wifi_client_->setCACert(baltimore_root_pem_);
+  } else {
+    ESP_LOGD(TAG, "Using insecure SSL");
+  }
+#endif
+  this->http_client_.setReuse(true);
+}
+
+std::string AzureIoTHub::get_iot_hub_device_id() const { return this->iot_hub_device_id_; }
+
+void AzureIoTHub::set_iot_hub_device_id(const std::string &device_id) { this->iot_hub_device_id_ = device_id; }
+void AzureIoTHub::set_iot_hub_sas_token(const std::string &sas_token) { this->iot_hub_sas_token_ = sas_token; }
+void AzureIoTHub::set_iot_hub_rest_url(const std::string &rest_url) { this->iot_hub_rest_url_ = rest_url; }
+void AzureIoTHub::set_iot_hub_sas_token_expiration_string(const std::string &expiration_string) {
+  this->iot_hub_sas_token_expiration_string_ = expiration_string;
+}
+
+#ifdef ARDUINO_ARCH_ESP8266
+void AzureIoTHub::set_iot_hub_ssl_sha1_fingerprint(const std::string &fingerprint) {
+  this->ssl_fingerprint_supplied_ = this->set_fingerprint_bytes_(fingerprint.c_str());
+}
+#endif
+
+bool AzureIoTHub::post_json_to_iot_hub_(const std::string json_payload) {
+  ESP_LOGD(TAG, "Posting to Azure IoT Hub");
+  String url{this->iot_hub_rest_url_.c_str()};
+
+#ifndef CLANG_TIDY
+  this->http_client_.begin(*this->wifi_client_, url);
+#endif
+
+  this->http_client_.addHeader(F("Authorization"), this->iot_hub_sas_token_.c_str());
+  this->http_client_.addHeader(F("Content-Type"), F("application/json"));
+  this->http_client_.addHeader(F("Content-Length"), String(json_payload.length()));
+
+  int http_code = this->http_client_.POST(json_payload.c_str());
+  this->http_client_.end();
+  ESP_LOGD(TAG, "HTTP Request to Azure IoT Hub completed with code: %d", http_code);
+  bool success = http_code == HTTP_CODE_OK || http_code == HTTP_CODE_NO_CONTENT;
+  if (!success) {
+    this->status_set_error();
+  } else if (this->status_has_error()) {
+    this->status_clear_error();
+  }
+
+  return success;
+}
+
+#ifdef USE_BINARY_SENSOR
+void AzureIoTHub::on_binary_sensor_update(binary_sensor::BinarySensor *sensor, bool state) {
+  if (sensor->is_internal())
+    return;
+
+  std::string payload = json::build_json([this, sensor, state](JsonObject &root) {
+    root[F("deviceId")] = this->get_iot_hub_device_id();
```

Comment:
```suggestion
    root["deviceId"] = this->get_iot_hub_device_id();
```

There's not much use in using `F()` for such short strings (the pointer itself is 4-byte long already).

Plus ESPHome doesn't have that much of an issue with program size as other firmwares by design.

Reply:
I am used to trying to optimise every single byte of runtime memory if I can. Thus on ESP8266 and ESP32 libraries that support PROGMEM (which ArduinoJson does), I tend to store all constant strings on flash. If this is a concern due to additional processing and time taken to take them out into runtime, I will amend. As you said - the saving here is minimal.